src/store/writer.rs

```diff
@@ -85,76 +129,137 @@ impl StoreWriter {
         VInt(doc_num_bytes as u64).serialize(&mut self.current_block)?;
         self.current_block
             .write_all(&self.intermediary_buffer[..])?;
-        self.doc += 1;
-        if self.current_block.len() > self.block_size {
-            self.write_and_compress_block()?;
-        }
+        self.num_docs_in_current_block += 1;
+        self.check_flush_block()?;
+        Ok(())
+    }
+
+    /// Store bytes of a serialized document.
+    ///
+    /// The document id is implicitely the current number
+    /// of documents.
+    pub fn store_bytes(&mut self, serialized_document: &[u8]) -> io::Result<()> {
+        let doc_num_bytes = serialized_document.len();
+        VInt(doc_num_bytes as u64).serialize(&mut self.current_block)?;
+        self.current_block.write_all(serialized_document)?;
+        self.num_docs_in_current_block += 1;
+        self.check_flush_block()?;
         Ok(())
     }
 
     /// Stacks a store reader on top of the documents written so far.
     /// This method is an optimization compared to iterating over the documents
     /// in the store and adding them one by one, as the store's data will
     /// not be decompressed and then recompressed.
-    pub fn stack(&mut self, store_reader: &StoreReader) -> io::Result<()> {
-        if !self.current_block.is_empty() {
-            self.write_and_compress_block()?;
-        }
-        assert_eq!(self.first_doc_in_block, self.doc);
-        let doc_shift = self.doc;
-        let start_shift = self.writer.written_bytes() as usize;
+    pub fn stack(&mut self, store_reader: StoreReader) -> io::Result<()> {
+        // We flush the current block first before stacking
+        self.send_current_block_to_compressor()?;
```

Comment:
Both errors are actually errors on the compressing thread.

If there is an error, could we join the companion thread and return 
- its io::Error if it has one
- a custome io::Error if it panicked.

The code that join/harvest the error could be factorized in an independant method.

Reply:
that's a good idea, I don't like the error handling here currently, it's not deterministic which error is returned .. but to join the thread we need to consume `self`. We could swap it with another handle or put it in an option, I don't really like either of those